source/routing/raptor_api.cpp

```diff
@@ -1368,73 +1369,79 @@ void filter_late_journeys(RAPTOR::Journeys& journeys, const NightBusFilter::Para
-            for (const auto& st :
-                 boost::make_iterator_range(vj->stop_time_list.begin() + order.val, vj->stop_time_list.end())) {
-                if (st.stop_point->stop_area->uri == stop_area_uri && vj != vj_to_skip) {
-                    return true;
-                }
-                if (section.get_out_st == &st) {
-                    reach_end = true;
-                    break;
+bool update_visited_section(navitia::routing::Journey::Section& section,
+                            const std::string& stop_area_uri,
+                            const navitia::type::VehicleJourney* vj_to_skip,
+                            const bool clockwise) {
+    log4cplus::Logger logger = log4cplus::Logger::getInstance(LOG4CPLUS_TEXT("logger"));
+    auto order = section.get_in_st->order();
+    // because of stay-ins, we may have several vj in one section, we have to scan the stop times
+    // of all vjs
+    for (const auto* vj = section.get_in_st->vehicle_journey; vj; (vj = vj->next_vj, order = type::RankStopTime(0))) {
+        for (const auto& st :
+             boost::make_iterator_range(vj->stop_time_list.begin() + order.val, vj->stop_time_list.end())) {
+            if (st.stop_point->stop_area->uri == stop_area_uri && vj != vj_to_skip) {
+                if (clockwise) {
+                    section.get_out_st = &st;
+                    section.get_out_dt = st.departure_time;
+                } else {
+                    section.get_in_st = &st;
+                    section.get_in_dt = st.arrival_time;
                 }
+                return true;
             }
-            if (reach_end) {
-                break;
+            if (section.get_out_st == &st) {
+                return false;
             }
         }
     }
+    LOG4CPLUS_WARN(logger, "Error occurred when modifying backtracking journeys");
     return false;
 }
 
-void filter_backtracking_journeys(RAPTOR::Journeys& journeys, const bool clockwise) {
-    log4cplus::Logger logger = log4cplus::Logger::getInstance(LOG4CPLUS_TEXT("logger"));
-
-    if (journeys.size() == 1) {
-        return;
+std::pair<bool, size_t> get_and_update_visited_section(navitia::routing::Journey& journey,
+                                                       const std::string& stop_area_uri,
+                                                       const navitia::type::VehicleJourney* vj_to_skip,
+                                                       const bool clockwise) {
+    // TODO: use boost::adaptors::indexed(0), index() and value() once boost>1.55 is available, sigh...
+    size_t section_idx = 0;
+    for (auto& section : journey.sections) {
+        if (update_visited_section(section, stop_area_uri, vj_to_skip, clockwise)) {
+            return {true, section_idx};
+        }
+        ++section_idx;
     }
-    auto it = journeys.begin();
-    while (it != journeys.end()) {
-        const auto& journey = *it;
+    return {false, 0};
```

Comment:
Why do you need to iterate through all sections ?
You know that you will modify either the last section (clockwise) or the first section (!clockwise)